ui/lib/kv/addon/components/page/secret/metadata/version-diff.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import { tracked } from '@glimmer/tracking';
+import { action } from '@ember/object';
+import { kvDataPath } from 'vault/utils/kv-path';
+import { typeOf } from '@ember/utils';
+
+/**
+ * @module KvVersionDiff
+ * This component produces a JSON diff view between 2 secret versions. It uses the library jsondiffpatch.
+ *
+ * @param {string} backend - Backend from the kv/data model.
+ * @param {string} path - Backend from the kv/data model.
+ * @param {array} metadata - The kv/metadata model. It is version agnostic.
+ * @param {array} breadcrumbs - Array to generate breadcrumbs, passed to the page header component.
+ */
+
+export default class KvVersionDiffComponent extends Component {
+  @service store;
+
+  @tracked leftSideVersion = this.args.metadata.currentVersion;
+  @tracked rightSideVersion = this.defaultRightSideVersion;
+  @tracked statesMatch = false;
+  @tracked visualDiff;
+
+  constructor() {
+    super(...arguments);
+    this.createVisualDiff();
+  }
+
+  get defaultRightSideVersion() {
+    // unless the version is destroyed or deleted we return the version prior to the current version.
+    const versionData = this.args.metadata.sortedVersions.find(
+      (version) =>
+        version.destroyed === false && version.deletion_time === '' && version.version != this.leftSideVersion
+    );
+    // if all versions have been deleted or destroyed return one less than the current version;
+    return versionData ? versionData.version : this.leftSideVersion - 1;
+  }
+
+  async fetchSecretData(version) {
+    // ARG TODO ask if maybe we want to override the adapter peekRecord?
```

Comment:
As per my comment in the code, originally Chelsea and I had talked about making a method in the adapter that did what I'm doing here. Checking the store and if nothing exists then hitting up queryRecord. We could do this by overriding peekRecord in the data adapter. 

Thoughts?